src/pubsub/src/publisher/worker.rs

```diff
@@ -152,57 +175,107 @@ impl Worker {
+            self.pending_batch.push(publish);
+            if self.at_batch_threshold() {
+                break;
+            }
         }
     }
 
-    pub(crate) fn flush(&mut self, inflight: &mut FuturesUnordered<tokio::task::JoinHandle<()>>) {
-        if !self.pending_batch.is_empty() {
-            inflight.push(
-                self.pending_batch
-                    .flush(self.client.clone(), self.topic.clone()),
-            );
+    /// The main loop of the batch worker.
+    ///
+    /// This method concurrently handles the following events:
+    ///
+    /// 1. A Publish command from the `Worker` causes the new message to be
+    ///    added a pending message queue. If there is enough message to create
+    ///    a full batch, then also flush the batch.
+    /// 2. A `Flush` command from the `Publisher` causes all the pending messages
+    ///    to be flushed respecting the configured batch size and message ordering.
+    /// 4. A `inflight` batch completion causes the next batch to send if it satisfies
+    ///    the configured batch threshold.
+    ///
+    /// The loop terminates when the `rx` channel is closed, which happens when the
+    /// `Worker` drops the Sender.
+    pub(crate) async fn run(mut self) {
+        // While it is possible to use Some(JoinHandle) here as there is at max
+        // a single inflight task at any given time, the use of JoinSet
+        // simplify the managing the inflight JoinHandle.
+        let mut inflight = JoinSet::new();
+        loop {
+            tokio::select! {
+                _ = inflight.join_next(), if !inflight.is_empty() => {
+                    self.move_to_batch();
+                    if self.at_batch_threshold() {
+                        self.pending_batch.flush(self.client.clone(), self.topic.clone(), &mut inflight);
+                    }
+                }
+                msg = self.rx.recv() => {
+                    match msg {
+                        Some(ToBatchWorker::Publish(msg)) => {
+                            self.pending_msges.push_back(msg);
+                            self.move_to_batch();
+                            if self.at_batch_threshold() && inflight.is_empty() {
+                                self.pending_batch.flush(self.client.clone(), self.topic.clone(), &mut inflight);
+                            }
+                        },
+                        Some(ToBatchWorker::Flush(tx)) => {
+                            if !inflight.is_empty() {
+                                inflight.join_next().await;
+                            }
+                            loop {
+                                if self.pending_batch.is_empty() && self.pending_msges.is_empty() {
+                                    break;
+                                }
+                                self.move_to_batch();
+                                self.pending_batch.flush(self.client.clone(), self.topic.clone(), &mut inflight);
```

Comment:
using flush for the timer seems slightly inefficient since if we have a max batch size of 1000 and we have a batch out and 1001 messages in the queue, then we we will send that single message as a batch of 1 even if more come in. I guess that is a possibility with having the timeout anyway, so we maybe its not a real problem and we don't need it to be changed necessarily.

This also makes we wonder actually if we have an outstanding batch, would it be better to just leave all messages in the channel? to avoid having to copy things around as much? I'm just curious about this. Something to maybe try out later.